### PR TITLE
feat:[CDS-52932]: Stage Validator Changes for Command and Rolling Steps

### PR DESCRIPTION
**Pull request by piyush-bhuwalka**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/43495)

### CODE REVIEW

**Comment by piyush-bhuwalka** [comment]
trigger feature-build

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/creator/plan/stage/TasStageValidatorHelper.java [hunk 1]
```diff
@@ -76,17 +79,25 @@ public void validate(Object object, String accountIdentifier, String orgIdentifi
...
     }
     int swapRouteStepCount = stepTypeToCount.getOrDefault(TAS_SWAP_ROUTES, 0);
     validateStepCount(swapRouteStepCount, "At max one Swap Route step is valid, found: %d");
     int appRollbackStepCount = stepTypeToCount.getOrDefault(TAS_ROLLBACK, 0);
     int swapRollbackStepCount = stepTypeToCount.getOrDefault(SWAP_ROLLBACK, 0);
+    int rollingRollbackStepCount = stepTypeToCount.getOrDefault(TAS_ROLLING_ROLLBACK, 0);
+    validateStepCount(rollingRollbackStepCount, "At max one Rolling Rollback step is valid, found: %d");
     validateStepCount(appRollbackStepCount, "At max one App Rollback step is valid, found: %d");
     validateStepCount(swapRollbackStepCount, "At max one Swap Rollback step is valid, found: %d");
     if ((appRollbackStepCount + swapRollbackStepCount) > 1) {
```
Comment (RishabhGupta34) [inline 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/creator/plan/stage/TasStageValidatorHelper.java:103]: this should have rolling as well

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/creator/plan/stage/TasStageValidatorHelper.java [hunk 2]
```diff
@@ -76,17 +79,25 @@ public void validate(Object object, String accountIdentifier, String orgIdentifi
...
     validateStepCount(canaryAppSetupStepCount, "Only one Canary App Setup step is valid, found: %d");
     validateStepCount(bgAppSetupStepCount, "Only one BG App Setup step is valid, found: %d");
-    if ((basicAppSetupStepCount + bgAppSetupStepCount + canaryAppSetupStepCount) != 1) {
+    validateStepCount(rollingDeployStepCount, "Only one Rolling Deploy step is valid, found: %d");
+
+    if (((basicAppSetupStepCount + bgAppSetupStepCount + canaryAppSetupStepCount) > 1)
+        || ((basicAppSetupStepCount + bgAppSetupStepCount + canaryAppSetupStepCount) == 0 && commandStepCount == 0
+            && rollingDeployStepCount == 0)) {
       throw new InvalidYamlException(format("Only one App Setup step out of %s is supported",
           List.of(TAS_BASIC_APP_SETUP, TAS_BG_APP_SETUP, TAS_CANARY_APP_SETUP)));
```
Comment (RishabhGupta34) [inline 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/creator/plan/stage/TasStageValidatorHelper.java:93]: can you change error message

**Comment by piyush-bhuwalka** [comment]
trigger feature-build

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/creator/plan/stage/TasStageValidatorHelper.java (hunk 1 above)
Comment (piyush-bhuwalka) [reply 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/creator/plan/stage/TasStageValidatorHelper.java:103]: Done

**Comment by piyush-bhuwalka** [comment]
trigger feature-build

**Comment by piyush-bhuwalka** [comment]
trigger feature-build

**Comment by piyush-bhuwalka** [comment]
trigger feature-build

**Review by sourabh-awashti — APPROVED** [review]
> Reviewed 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43495)* status: 2 file(s) to review. 1. unresolved discussions. 1 of 2 approvals obtained.

**Review by RishabhGupta34 — APPROVED** [review]

**Review by RishabhGupta34 — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43495#-:-NNad8Kz74jpkNTusD12:bnfp4nl)* status: 1. unresolved discussions.

**Comment by piyush-bhuwalka** [comment]
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/creator/plan/stage/TasStageValidatorHelper.java` line 93 at r1](https://reviewable.io/reviews/harness/harness-core/43495#-NNM5jQfmmagf9R7H2Uj-r1-93:-NNahp9y4m970Tt9Ls9w:bb7f2mf) ([raw file](https://github.com/harness/harness-core/blob/43eda235bf205e3506200c5892f6c008bd1effab/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/creator/plan/stage/TasStageValidatorHelper.java#L93)):*
<details><summary><i>Previously, RishabhGupta34 (Rishabh Gupta) wrote…</i></summary><blockquote>

can you change error message
</blockquote></details>

done

**Review by RishabhGupta34 — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43495)* status: 1. unresolved discussions.

**Review by RishabhGupta34 — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43495)* status: :shipit: complete! ✓

**Pull request merged**